microsoft/testsuites/power/power.py

```diff
@@ -0,0 +1,587 @@
+
+        dmesg = node.tools[Dmesg]
+        root_msg = re.findall('root=', dmesg.get_output(force_run=True), re.IGNORECASE)
+        env_setup_result = [res for res in root_msg if 'resume' in res]
+        log.debug("Env Setup Result: " + ''.join(env_setup_result))
+        if env_setup_result:
+            log.info("Evironment has been set up...")
+        else:
+            # Configuration for hibernation
+            node.execute(
+                "umask 022"
+            )
+
+            # GetGuestGeneration
+            exit_code = node.execute(
+                "ls /sys/firmware/efi/"
+            ).exit_code
+            os_generation = 2 if exit_code == 0 else 1
+            
+            if isinstance(node.os, Redhat):
+                min_support_kernel = "4.18.0-202"
+                check_vm_support_res = check_vm_feature_support_status(min_support_kernel)
+                current_kernel_version = node.execute(
+                    "uname -r", shell = True, sudo = True
+                )
+                assert_that(
+                    check_vm_support_res,
+                    f"Hibernation is supported since kernel-4.18.0-202. Current version: {current_kernel_version}. Skip the test."
+                ).is_true()
+            
+            # prepare swap space
+            # Get the latest device name which should be the new attached data disk
+            # TODO: USE FDISK TOOL
+            data_dev = node.execute(
+                "ls /dev/sd*[a-z] | sort -r | head -1", shell = True, sudo = True
+            ).stdout
+            fdisk = node.tools[Fdisk]
+            # 82  Linux swap / So
+            fdisk.change_partition_type(data_dev, '82')
+            log.debug("Executed fdisk command to change partition type to swap.")
+        
+            cmd_res = node.execute(
+                "ls /dev/sd*", shell = True, sudo = True
+            )
+
+            swap_partition = data_dev + '1'
+            node.tools[Swap].set_up_swap_space(swap_partition)
+
+            node.tools[SwapOn].run(swap_partition, shell=True, sudo=True)
+            node.tools[SwapOn].run("-s", shell=True, sudo=True)
+
+            # Seems there may not be 'sw' in blkid, so use partition name to check
+            partition_info = node.tools[Blkid].get_partition_info_by_name(swap_partition)
+            sw_uuid = "UUID=" + partition_info.uuid
+            log.debug(f"Found the Swap space disk UUID: {sw_uuid}")
+            if not sw_uuid:
+                raise SkippedException("Swap space disk UUID is empty. Abort the test.")
+
+            # Config the swap partition in /etc/fstab
+            # So it will be mounted automatically during reboot
+            node.execute(
+                "chmod 766 /etc/fstab", shell = True, sudo = True
+            )
+            node.tools[Echo].write_to_file(
+                f"{sw_uuid} none swap sw 0 0",
+                "/etc/fstab",
+                sudo = True,
+                append = True
+            )
```

Comment:
need recover /etc/fstab it after testing

Reply:
do we also need to restore other files? e.g. /etc/default/grub.d/50-cloudimg-settings.cfg

Reply:
It's better to. Make sure the recovered environment is still working.